compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,449 @@
+                             const luci::CircleTranspose *b_tr);
+
+private:
+  void init_shape_with_tag(const luci::CircleNode *in_tensor);
+
+  template <loco::DataType PermType>
+  bool analyze_transpose(const luci::CircleTranspose *transpose_node);
+
+  template <loco::DataType ShapeType> bool analyze_reshape(const luci::CircleReshape *reshape_node);
+
+  bool verify_tag() const;
+
+  struct Dim final
+  {
+    uint32_t value;
+    std::vector<uint8_t> tags;
+  };
+  using Shape = std::vector<Dim>;
+
+  const uint8_t START_TAG = 0;
+  Shape _shape;
+};
+
+/**
+ * @brief initalize _shape with input tensor named in_tensor
+ *
+ * @note 'tags' are attached to non-1 valued dimension.
+ */
+void TaggedShapeAnalyzer::init_shape_with_tag(const luci::CircleNode *in_tensor)
+{
+  _shape.clear();
+  uint8_t tag = START_TAG;
+
+  for (uint32_t i = 0; i < in_tensor->rank(); i++)
+  {
+    TaggedShapeAnalyzer::Dim dim;
+    {
+      dim.value = in_tensor->dim(i).value();
+      if (dim.value != 1)
+        dim.tags.push_back(tag++);
+    }
+    _shape.push_back(dim);
+  }
+}
+
+/**
+ * @brief update _shape based on 'Transpose' permutation value
+ *
+ * @example Let's assume Transpose(perm=0, 3, 1, 2) is given to [before] _shape.
+ *
+ *  This function reordered the Dims' order based on permutaiton value.
+ *
+ *  [before]  _shape :
+ *              - value :   (1)   (7)   (7)   (448)
+ *              - tags  :   (-)   (0)   (1)   (2)
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (7)  (7)
+ *              - tags  :   (-)   (2)    (0)  (1)
+ */
+template <loco::DataType PermType>
+bool TaggedShapeAnalyzer::analyze_transpose(const luci::CircleTranspose *transpose_node)
+{
+  const luci::CircleConst *perm_node = loco::must_cast<luci::CircleConst *>(transpose_node->perm());
+  assert(perm_node->dtype() == PermType);
+
+  TaggedShapeAnalyzer::Shape new_shape;
+  const auto size = perm_node->size<PermType>();
+  for (uint32_t i = 0; i < size; i++)
+  {
+    auto perm_idx = perm_node->at<PermType>(i);
+    new_shape.push_back(_shape.at(perm_idx));
+  }
+  _shape = new_shape;
+  return true;
+}
+
+/**
+ * @brief update _shape based on 'Reshape' shape value
+ *
+ * @example Let's assume Reshape(shape=1, 448, 49) is given to [before] _shape.
+ *
+ *  [before]  _shape :
+ *              - value :   (1)   (448)  (7)   (7)
+ *              - tags  :   (-)   (2)    (0)   (1)
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (49)
+ *              - tags  :   (-)   (2)    (0, 1)
+ */
+template <loco::DataType ReshapeType>
+bool TaggedShapeAnalyzer::analyze_reshape(const luci::CircleReshape *reshape_node)
+{
+  const luci::CircleConst *shape_node = loco::must_cast<luci::CircleConst *>(reshape_node->shape());
+  assert(shape_node->dtype() == ReshapeType);
+
+  // At least one element must be in reshape's output-tensor.
+  if (shape_node->size<ReshapeType>() <= 0)
+    return false;
```

Comment:
Does this case happen?

Reply:
I don't think so. 
But I'm not 100% sure.
 
I don't think the caller could guarantee that `reshape/shape` has at least one element. 
So, That's why I explicitly handle this case, instead of using assertion. 

----------------------------

Do you think it is better to use `assertion` here? 
If so, plz let me know.

Reply:
Hmm.. well, both way looks good to me. Let's go as is.